shared/src/balancer_event_handler.rs

```diff
@@ -7,31 +7,107 @@ use crate::{
 use anyhow::{anyhow, Context, Result};
 use contracts::{
     balancer_v2_vault::{
-        self, event_data::PoolRegistered as ContractPoolRegistered, Event as ContractEvent,
+        self,
+        event_data::{
+            PoolRegistered as ContractPoolRegistered, TokensRegistered as ContractTokensRegistered,
+        },
+        Event as ContractEvent,
     },
     BalancerV2Vault,
 };
 use ethcontract::common::DeploymentInformation;
 use ethcontract::{dyns::DynWeb3, Event as EthContractEvent, EventMetadata, H160, H256};
-use std::fmt::Debug;
-use std::ops::RangeInclusive;
+use std::{
+    collections::{HashMap, HashSet},
+    fmt::Debug,
+    ops::RangeInclusive,
+};
 use tokio::sync::Mutex;
 
-#[derive(Debug)]
+#[derive(Clone, Debug)]
 pub enum BalancerEvent {
     PoolRegistered(PoolRegistered),
+    TokensRegistered(TokensRegistered),
 }
 
-#[derive(Debug)]
+#[derive(Debug, Clone)]
 pub struct PoolRegistered {
     pub pool_id: H256,
     pub pool_address: H160,
     pub specialization: PoolSpecialization,
 }
 
+#[derive(Debug, Clone)]
+pub struct TokensRegistered {
+    pub pool_id: H256,
+    pub tokens: Vec<H160>,
+}
+
+#[derive(Clone, Debug, Default, Eq, PartialEq, Hash)]
+pub struct WeightedPool {
+    pool_id: H256,
+    pool_address: Option<H160>,
+    tokens: Option<Vec<H160>>,
+    specialization: Option<PoolSpecialization>,
+    block_created: u64,
+}
+
+impl WeightedPool {
+    fn update_from_event(&mut self, event: BalancerEvent) {
+        // Pool and Token Registration always occur in the same tx! So block_created doesn't need update.
+        // https://github.com/balancer-labs/balancer-v2-monorepo/blob/70843e6a61ad11208c1cfabf5cfe15be216ca8d3/pkg/pool-utils/contracts/BasePool.sol#L128-L130
+        match event {
+            BalancerEvent::PoolRegistered(pool_registration) => {
+                self.pool_address = Some(pool_registration.pool_address);
+                self.specialization = Some(pool_registration.specialization);
+            }
+            BalancerEvent::TokensRegistered(token_registration) => {
+                self.tokens = Some(token_registration.tokens)
+            }
+        }
+    }
+}
+
+impl From<(EventIndex, PoolRegistered)> for WeightedPool {
```

Comment:
nanonit: Maybe you can just get away with a single new and then use `update_from_event` for both the  `PoolRegistered` and `TokensRegistered`

Reply:
I have tried a few alternatives to this, but none of them seem to turn out very well.

Example:

```rust
fn new(index: EventIndex, pool_registration: Option<PoolRegistration>,  token_registration: Option<TokenRegistration>) -> Self
```
Would be cool if I could override the new method with an implementation for each of the above options.